src/Rel8/Tabulate.hs

```diff
@@ -0,0 +1,616 @@
+import Rel8.Expr ( Expr )
+import Rel8.Expr.Aggregate ( countStar )
+import Rel8.Expr.Bool ( true )
+import Rel8.Order ( Order( Order ) )
+import Rel8.Query ( Query )
+import qualified Rel8.Query.Exists as Q ( exists, present, absent )
+import Rel8.Query.Filter ( where_ )
+import Rel8.Query.List ( catNonEmptyTable )
+import qualified Rel8.Query.Maybe as Q ( optional )
+import Rel8.Query.Opaleye ( mapOpaleye )
+import Rel8.Query.These ( alignBy )
+import Rel8.Table ( Table, fromColumns, toColumns )
+import Rel8.Table.Aggregate ( hgroupBy, listAgg, nonEmptyAgg )
+import Rel8.Table.Alternative
+  ( AltTable, (<|>:)
+  , AlternativeTable, emptyTable
+  )
+import Rel8.Table.Cols ( fromCols, toCols )
+import Rel8.Table.Eq ( EqTable, (==:), eqTable )
+import Rel8.Table.List ( ListTable( ListTable ) )
+import Rel8.Table.Maybe ( MaybeTable( MaybeTable ), maybeTable )
+import Rel8.Table.NonEmpty ( NonEmptyTable( NonEmptyTable ) )
+import Rel8.Table.Opaleye ( aggregator, unpackspec )
+import Rel8.Table.Ord ( OrdTable )
+import Rel8.Table.Order ( ascTable )
+import Rel8.Table.Projection
+  ( Biprojectable, biproject
+  , Projectable, project
+  , apply
+  )
+import Rel8.Table.These ( TheseTable( TheseTable ), theseTable )
+
+-- semigroupoids
+import Data.Functor.Apply ( Apply, liftF2 )
+import Data.Functor.Bind ( Bind, (>>-) )
+
+
+type Key :: Type -> Type
+type Key = Maybe
+
+
+cat :: Table Expr k => Key k -> Query k
+cat = maybe emptyTable pure
+
+
+key :: (ProductProfunctor p, SumProfunctor p)
+  => p a b -> p (Key a) (Key b)
+key a = dimap from to (PP.empty +++! a)
+  where
+    from = maybe (Left ()) Right
+    to = either (const Nothing) Just
+
+
+keyed :: (ProductProfunctor p, SumProfunctor p)
+  => p k l -> p a b -> p (Key k, a) (Key l, b)
+keyed k a = key k ***! a
+
+
+type Predicate :: Type -> Type
+newtype Predicate a = Predicate (Maybe (a -> Expr Bool))
+
+
+instance Contravariant Predicate where
+  contramap f (Predicate a) = Predicate (lmap f <$> a)
+
+
+instance Semigroup (Predicate k) where
+  Predicate ma <> Predicate mb = Predicate $ ma <|> mb
+
+
+instance Monoid (Predicate k) where
+  mempty = Predicate Nothing
+
+
+match :: EqTable k => Key k -> Predicate k
+match = Predicate . fmap (==:)
+
+
+ensure :: Predicate k -> Key k -> Query ()
+ensure (Predicate mp) = traverse_ (\k -> traverse_ (\p -> where_ (p k)) mp)
+
+
+-- | A @'Tabulation' k a@ is like a @'Query' a@, except that each row also
+-- has a key @k@ in addition to the value @a@. 'Tabulation's can be composed
+-- monadically just like 'Query's, but the resulting join is more like a
+-- @NATURAL JOIN@ (based on the common key column(s) @k@) than the
+-- @CROSS JOIN@ given by 'Query'.
+--
+-- Another way to think of @'Tabulation' k a@ is as analogous to @Map k a@ in
+-- the same way @'Query' a@ is analogous to @[a]@. However, there's nothing
+-- stopping a 'Tabulation' from containing multiple rows with the same key, so
+-- technically @Map k (NonEmpty a)@ is more accurate.
```

Comment:
I think it'd be good to say something about the introduction and elimination of `Tabulation`s - maybe just mention "you can create `Tabulation`s with `fromQuery` or `liftQuery`, and convert them back into `Query`s with `toQuery` and `lookup`"?